File: 백준/Gold/7569. 토마토/토마토.py
from collections import deque

M, N, H = map(int, input().split())

arr = [[list(map(int, input().split())) for _ in range(N)] for __ in range(H)]
visited = [[[0 for _ in range(M)] for _ in range(N)] for _ in range(H)]

def bfs():
    cnt = N * M * H
    q = deque()

    for i in range(H):
        for j in range(N):
            for k in range(M):
                if arr[i][j][k] == 1:
                    q.append((i, j, k))
                    visited[i][j][k] = 1
                    cnt -= 1
                elif arr[i][j][k] == -1:
                    cnt -= 1

    if cnt == 0:
        return 0

    while q:
        i, j, k = q.popleft()

        for di, dj, dk in [(1, 0, 0), (-1, 0, 0), (0, 1, 0), (0, -1, 0), (0, 0, 1), (0, 0, -1)]:
            ni, nj, nk = i + di, j + dj, k + dk
            if 0 <= ni < H and 0 <= nj < N and 0 <= nk < M:
                if visited[ni][nj][nk] == 0 and arr[ni][nj][nk] == 0:
                    visited[ni][nj][nk] = visited[i][j][k] + 1
                    arr[ni][nj][nk] = 1
                    q.append((ni, nj, nk))
                    cnt -= 1

    if cnt > 0:
        return -1

    ans = 0
    for i in range(H):
        for j in range(N):
            for k in range(M):
                ans = max(ans, visited[i][j][k])

    return ans - 1

print(bfs())

File: 백준/Gold/7569. 토마토/토마토.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int M, N, H;
int dz[6] = { 0,0,0,0,1,-1 };
int dy[6] = { 0,0,1,-1,0,0 };
int dx[6] = { 1,-1,0,0,0,0 };

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> M >> N >> H;
	vector<vector<vector<int>>> arr(H, vector<vector<int>>(N, vector<int>(M, 0)));
	vector<vector<vector<int>>> visited(H, vector<vector<int>>(N, vector<int>(M, 0)));
	queue<vector<int>> q;

	for (int h = 0; h < H; h++) {
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				cin >> arr[h][i][j];
				if (arr[h][i][j] == 1) {
					visited[h][i][j] = 1;
					q.push({ h, i, j });
				}
			}
		}
	}

	while (!q.empty()) {
		int z = q.front()[0];
		int y = q.front()[1];
		int x = q.front()[2];
		q.pop();

		for (int d = 0; d < 6; d++) {
			int nz = z + dz[d];
			int ny = y + dy[d];
			int nx = x + dx[d];

			if (nz < 0 || ny < 0 || nx < 0 || nz >= H || ny >= N || nx >= M) continue;
			if (visited[nz][ny][nx] == 0 && arr[nz][ny][nx] == 0) {
				visited[nz][ny][nx] = visited[z][y][x] + 1;
				q.push({ nz, ny, nx });
			}
		}
	}

	int ans = 0;
	bool bp = false;
	for (int h = 0; h < H; h++) {
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if (arr[h][i][j] != -1 && visited[h][i][j] == 0) {
					ans = -1;
					bp = true;
					break;
				}
				ans = max(ans, visited[h][i][j] - 1);
			}
			if (bp) break;
		}
		if (bp) break;
	}

	cout << ans;
	return 0;
}
